src/source/custom_source.js

```diff
@@ -323,7 +326,17 @@ class CustomSource<T> extends Evented implements Source {
 
         const {x, y, z} = tile.tileID.canonical;
         const data = this._implementation.prepareTile({x, y, z});
-        if (!data) return null;
+
+        // If the implementation returned `undefined` as tile data, return immideatly to indicate that we have no new data for the tile.
+        // A map will render a previous version of the tile in the tile’s space if there was any.
+        if (data === undefined) return null;
+
+        // If the implementation returned `null` as tile data, mark the tile as `loading`.
+        // A SourceCache will trigger the tile loading. A map will render nothing in the tile’s space.
+        if (data === null) {
+            tile.state = 'loading';
```

Comment:
Why should `null` mark the tile as `loading` not `loaded`?

Reply:
SourceCache will not trigger tile reloading since it's `loaded` and will use a previous version of the tile. However, if the tile state is set to `loading`, SourceCache will render nothing instead of a previous version of a tile and trigger tile reloading.

If we use `reloading` as a tile state, SourceCache will trigger tile reloading, but it would also render the previous version of the tile.

https://github.com/mapbox/mapbox-gl-js/blob/cd9f65c796e09d92e4237f6afa5f5dbc71fc05d7/src/source/tile.js#L63-L70